### DISCUSSION TITLE
Allow duplicating flows

### DISCUSSION THREAD

**Discussion by DeanKamali** [directus/directus#14405] [Feature Requests]
Hey everyone, 

Would be nice to be able to copy / clone an existing flow so I can modify it slightly to perform another business logic. 

Thanks

**Comment by rijkvanzanten** [comment]
The short answer to _"Why can't I clone / copy an existing flow?"_ is because we haven't implemented it! Lemme convert this to the feature request for that 🙂

**Comment by pstuifzand** [comment]
Duplicating a flow would also be useful so you can keep the current flow running and base a new flow on it to for testing. When the new flow is ready you can replace the old flow.

**Reply by benhaynes** [reply]
When duplicating, the new flow should automatically be set to inactive to avoid duplicate runs.

**Comment by qbdanalytics** [comment]
Meanwhile this feature is been developed.
Is there a workaround to do it via API ?

**Reply by rijkvanzanten** [reply]
You can use the [Flows](https://docs.directus.io/reference/system/flows.html) and [Operations](https://docs.directus.io/reference/system/operations.html) APIs to modify them 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by M-GRV** [comment]
Yes please! Maybe while tackling this, add UI for import/export on the flows page ;)

https://github.com/directus/directus/discussions/17484#discussioncomment-4938677

**Reply by nico-k3** [reply]
Importing / exporting Flows would be really nice to allow for a quick setup of new client cms's. A lot of stuff gets used on multiple sites like automatic slug creation, form submission or search indexing. Currently a little tedious to setup every time.

**Comment by rijkvanzanten** [comment]
@david-zacharias did a great RFC to get this moving forwards in #18559:

---

### Summary

Allow users to duplicate flows via Directus Studio UI

### Basic Example

_No response_

### Motivation

Users may want to try out changes to a flow but want to keep the original one as a backup.

### Detailed Design

### UI

In alignment to duplication of fields, the duplication can be triggered via the Select-Menu of each Flow in the List of Flows:

<img width="250" alt="Bildschirmfoto 2023-05-11 um 15 44 58" src="https://github.com/directus/directus/assets/1903179/e5337c95-cf29-4efb-9fb4-5927e98e1032">

After clicking a modal should be opened, asking the user to name the new flow. As default the flows current name appended with `(Copy)` is prefilled.

Users must confirm which leads to flow duplication or cancel.

### UX

The new flow should be disabled by default, to not have same flows being active in parallel.

### Logic

A flow consits of one `flow` record and 0 to n `operation` records, wich are linked via relations:

Relational properties of a **Flows**:

| Property  | Note |
| ---- | ---- |
| `operation` | UUID of the operation connected to the trigger in the flow. |


Relational properties an **Operation**:

| Property | Contains |
| ---- | ---- |
|`resolve` | The operation triggered when the current operation succeeds (or then logic of a condition [operation](https://docs.directus.io/reference/system/operations.html#the-operation-object)). Primary key of an operation. |
|`reject` | The operation triggered when the current operation fails (or otherwise logic of a condition [operation](https://docs.directus.io/reference/system/operations.html#the-operation-object)). Primary key of an operation. |
| `flow` | The flow containing this operation. Many-to-one to [flows](https://docs.directus.io/reference/system/flows.html#the-flow-object). |

The current implementation of the Directus Studio UI for managing flows uses a PATCH request on `flows/<the-flow-uuid>` with nested data like:

```json
{
  "name": "Some flow",
  "operation": { 
    "id": "cdd92780-2c55-4a52-bb1a-9f80a3d55900",
    "flow": "1c797550-ff17-49ff-93e8-2dbe4f3b8931",
    "resolve": {
      "id": "8fb91151-07b7-4e09-9e04-bbdb86f85ea7"
      "flow": "1c797550-ff17-49ff-93e8-2dbe4f3b8931"
      "reject": null
    }
  }
}
```

This allows to change / add the flow and all operations in one transaction.

Data from the backend (`flows/<uuid>?fields=*,operations.*`) is received as flat list of operations with references (*). We must transpose this list to a a tree like structure with `flow.operation` as root node then further spanned via `operation.resolve` and `operation.reject`. From all nodes (including the `flow` itself) the `id` property must be excluded and `flow`-property from each operation. 

To disable the duplicated flow by default `flow.status` should be set to `inactive`.

This new structure must be posted to `/flows` which then recursively creates all operations including new uuids for each operation. 


#### Edge case
There are may dangling operations which are configured but not linked to another operation or the flow (`flow.operation`) itself. These will not be included in the structured data described before, and must be added separately. Maybe via `flow.operations`-relational-property


(*) We could make use `?field`-Query  to receive nested relational field data, but this is probably less performant. And we must know / compute the maximum depth of the operations tree in advance.

### Requirements List

### Must Have:
- A Flow including all operations must be copied

### Should Have:
- C Possibility to give the copied flow a new name before advance of duplication

### Could Have:
- D

### Won't Have:
- E


### Drawbacks

I see no drawbacks, the feature is something users presumably feel natural to exist and can use without additional learning.

### Alternatives

No alternatives considered 

### Adoption Strategy

Neither data model nor existing business logic is altered; no breaking changes.

### Unresolved Questions

_No response_

**Comment by w0kyj** [comment]
Updates from Internal Docs:
# Motivation

- Save time when working with flows
- Administrators often need similar operations across different actions, workflows & triggers
- It solves the issue of having to remember the way a flow was setup.
- If building a similar flow to a already existing flow, this solves the problem of having to manually copy operations from the existing flow.
- Users may want to try out changes to a flow but want to keep the original one as a backup.
    - Backup existing flow before further enhancements are made
    - Use existing flow as starting point for similar operations or more complex requirements
- Benefits - Directus system admins

What is the expected outcome?

- A new flow with a copy of the existing flow’s setup and config. The user will be prompted to provide a new name as way of differentiating the existing and new flows and may allow override of default status `Inactive`.

# Detailed Design
...see comment above

## Requirements List

### Must haves:

- [ ]  ‘Copy flow’ button within the flow context menu
- [ ]  Modal to provide a new name to the copied flow
- [ ]  Flow including all operations must be copied

### Should haves:

- [ ] Auto/Default duplicated flow as `Inactive` (prevent same flow logic executing multiple times)

### Could haves:

- [ ]  Track original flow id and copy date
    - [ ]  Improved UI to display flows and nested flows based on where they were copied from. (e.g. add columns to the flows view)
- [ ]  Ability to compare flows
- [ ]  Option to include logs when a flow is copied
    - [ ] This is optional and perhaps limited to last "n" records

### Will not have:

- [ ] Upgrading to Table Layout and Search/Filter capabilities

# Drawbacks

- Potential risk that the list of flows will become confusing when many copies exist.
- Potential risk that duplicate flows are enabled by accident, causing unwanted behavior.

**Reply by Pachat** [reply]
I see only advantages to this feature. The disadvantages mentioned do not seem specific. Renaming the copied flow should be enough to maintain order. Should the operations themselves be renamed?

**Comment by geekyayush** [comment]
Any updates on this?

**Comment by kunalkeshan** [comment]
Is this still a work in progress?

**Comment by JayShoe** [comment]
There is a really nice extension that has this functionality already. 
https://github.com/baguse/directus-extension-flow-manager

**Reply by baguse** [reply]
Ahaha Thanks for mentioning it @JayShoe

**Comment by JamesW1** [comment]
Thanks for the suggestion! We've moved feature requests to our public roadmap at https://roadmap.directus.io/roadmap. 

I've migrated this request there for you: https://roadmap.directus.io/roadmap?id=0f87ab46-2ded-4ebd-9b38-57d4022f0ae8
